### PR TITLE
Disk quotas should not break if the rootfs contains the vcap user

### PR DESCRIPTION
**Pull request by badie**
[#95694242]

Since we're removing the creation of vcap from Garden, the vcap user needs to be added to the CF rootfs. Since the same rootfs will be used with both Garden and Warden, Warden needs to be able to handle vcap already existing. Currently it won't try to recreate vcap if it already exists, but the problem is that then all vcap users across containers will share the same UID, which will break disk quotas. Change the UIDs to make sure disk quotas still work.

Signed-off-by: Afolabi Badmos <abadmos@pivotal.io>

### CODE REVIEW

File: warden/spec/container/linux_spec.rb [hunk 1]
```diff
@@ -317,21 +317,112 @@ def self.it_should_stop_container_when_an_oom_event_occurs
...
+        limit_response.byte_limit.should == 2 * 1024 * 1024
+
+        vcap_limit_response = limit_disk(:handle => vcap_handle)
+        vcap_limit_response.byte_limit.should == 4 * 1024 * 1024
+
+        response = run("dd if=/dev/zero of=/tmp/test bs=1M count=3", vcap_handle)
+        response.exit_status.should == 0
+      end
+
+      it 'should still own his files' do
```
Comment (goonzoid) [inline warden/spec/container/linux_spec.rb]: I think it would be better to reword this to be gender neutral, e.g. `should still own vcap's files`.

Please can we change this?

Comment (badie) [inline warden/spec/container/linux_spec.rb]: It has been updated.

Comment (goonzoid) [inline warden/spec/container/linux_spec.rb]: Thanks! :+1:

**Comment by zaksoup** [comment]
Hi @badie 

We're having trouble getting a rootfs with the vcap user already existing to work. We see errors from the DEA and from warden.

From the dea:
```json
{"timestamp":1433973045.8336666,"message":"instance.start.failed with error Script exited with status 2","log_level":"warn","source":"Dea::Instance","data":{"attributes":{"stack":"cflinuxfs2","prod":false,"executableFile":"deprecated","limits":{"mem":256,"disk":1024,"fds":16384},"cc_partition":"default","console":false,"debug":null,"start_command":null,"health_check_timeout":null,"vcap_application":{"limits":{"mem":256,"disk":1024,"fds":16384},"application_version":"749a7e33-8210-42cd-ae71-3998ee374d3b","application_name":"4c2cc527-5b1a-4ce3-78f0-8a7ed5f80327","application_uris":["4c2cc527-5b1a-4ce3-78f0-8a7ed5f80327.10.244.0.34.xip.io"],"version":"749a7e33-8210-42cd-ae71-3998ee374d3b","name":"4c2cc527-5b1a-4ce3-78f0-8a7ed5f80327","space_name":"CATS-SPACE-3-2015_06_10-14h33m15.5s","space_id":"5f7a8805-951a-4d6d-8687-228b8c8ab730","uris":["4c2cc527-5b1a-4ce3-78f0-8a7ed5f80327.10.244.0.34.xip.io"],"users":null},"egress_network_rules":[{"destination":"0.0.0.0/0","ports":"53","protocol":"tcp"},{"destination":"0.0.0.0/0","ports":"53","protocol":"udp"},{"destination":"10.244.0.34","protocol":"all"},{"destination":"0.0.0.0-9.255.255.255","protocol":"all"},{"destination":"11.0.0.0-169.253.255.255","protocol":"all"},{"destination":"169.255.0.0-172.15.255.255","protocol":"all"},{"destination":"172.32.0.0-192.167.255.255","protocol":"all"},{"destination":"192.169.0.0-255.255.255.255","protocol":"all"},{"destination":"10.244.1.0/24","protocol":"all"},{"destination":"10.244.3.0/24","protocol":"all"}],"instance_index":0,"application_version":"749a7e33-8210-42cd-ae71-3998ee374d3b","application_name":"4c2cc527-5b1a-4ce3-78f0-8a7ed5f80327","application_uris":["4c2cc527-5b1a-4ce3-78f0-8a7ed5f80327.10.244.0.34.xip.io"],"application_id":"8b434108-13ed-4f25-a2af-d544f1e65ee5","droplet_sha1":"ec25751d6a0729d59f494ee84b134d74ff603376","instance_id":"fde9d4174e9548608d0d62cfd7735947","private_instance_id":"76b796ead53047e09cb8d58c87b295df2cca7493f6724a0b9d59d4d88bac8aa1","state":"CRASHED","state_timestamp":1433973045.8307836,"state_born_timestamp":1433973044.6702342,"state_starting_timestamp":1433973044.6719089,"warden_handle":"18o62fv0gjr","state_crashed_timestamp":1433973045.8307881},"duration":1.159351885,"error":"Script exited with status 2","backtrace":["/var/vcap/packages/dea_next/lib/container/container.rb:81:in `run_script'","/var/vcap/packages/dea_next/lib/dea/starting/instance.rb:427:in `block in promise_extract_droplet'","/var/vcap/packages/dea_next/lib/dea/promise.rb:92:in `call'","/var/vcap/packages/dea_next/lib/dea/promise.rb:92:in `block in run'"]},"thread_id":70101165527800,"fiber_id":70101202324420,"process_id":28791,"file":"/var/vcap/packages/dea_next/lib/dea/task.rb","lineno":104,"method":"block in resolve_and_log"}
{"timestamp":1433973049.7425227,"message":"\"cd /home/vcap/ && tar zxf /var/vcap/data/dea_next/droplets/a68acade6a4c8f82ae5c4cc787c2a65fae26a731/droplet.tgz\" exited with status 2 with data {:script=>\"cd /home/vcap/ && tar zxf /var/vcap/data/dea_next/droplets/a68acade6a4c8f82ae5c4cc787c2a65fae26a731/droplet.tgz\", :exit_status=>2, :stdout=>\"\", :stderr=>\"tar: ./logs: Cannot mkdir: Permission denied\\ntar: ./logs: Cannot mkdir: Permission denied\\ntar: ./logs/staging_task.log: Cannot open: No such file or directory\\ntar: ./tmp: Cannot mkdir: Permission denied\\ntar: ./staging_info.yml: Cannot open: Permission denied\\ntar: .: Cannot utime: Operation not permitted\\ntar: Exiting with failure status due to previous errors\\n\"}","log_level":"warn","source":"Container","data":{},"thread_id":70101165527800,"fiber_id":70101200315400,"process_id":28791,"file":"/var/vcap/packages/dea_next/lib/container/container.rb","lineno":80,"method":"run_script"}
{"timestamp":1433973049.746485,"message":"instance.start.failed with error Script exited with status 2","log_level":"warn","source":"Dea::Instance","data":{"attributes":{"stack":"cflinuxfs2","prod":false,"executableFile":"deprecated","limits":{"mem":256,"disk":1024,"fds":16384},"cc_partition":"default","console":false,"debug":null,"start_command":null,"health_check_timeout":null,"vcap_application":{"limits":{"mem":256,"disk":1024,"fds":16384},"application_version":"dc2a9548-6f5a-4e5f-8de2-00a8abac81ff","application_name":"225e916e-662b-4da9-41c0-712dbba1ccb3","application_uris":["225e916e-662b-4da9-41c0-712dbba1ccb3.10.244.0.34.xip.io"],"version":"dc2a9548-6f5a-4e5f-8de2-00a8abac81ff","name":"225e916e-662b-4da9-41c0-712dbba1ccb3","space_name":"CATS-SPACE-2-2015_06_10-14h33m15.507s","space_id":"dddc7196-7084-4632-8f6c-435a524b4892","uris":["225e916e-662b-4da9-41c0-712dbba1ccb3.10.244.0.34.xip.io"],"users":null},"egress_network_rules":[{"destination":"0.0.0.0/0","ports":"53","protocol":"tcp"},{"destination":"0.0.0.0/0","ports":"53","protocol":"udp"},{"destination":"10.244.0.34","protocol":"all"},{"destination":"0.0.0.0-9.255.255.255","protocol":"all"},{"destination":"11.0.0.0-169.253.255.255","protocol":"all"},{"destination":"169.255.0.0-172.15.255.255","protocol":"all"},{"destination":"172.32.0.0-192.167.255.255","protocol":"all"},{"destination":"192.169.0.0-255.255.255.255","protocol":"all"},{"destination":"10.244.1.0/24","protocol":"all"},{"destination":"10.244.3.0/24","protocol":"all"}],"instance_index":0,"application_version":"dc2a9548-6f5a-4e5f-8de2-00a8abac81ff","application_name":"225e916e-662b-4da9-41c0-712dbba1ccb3","application_uris":["225e916e-662b-4da9-41c0-712dbba1ccb3.10.244.0.34.xip.io"],"application_id":"5985cb75-53f1-4ec9-89eb-d068f1e8f02d","droplet_sha1":"a68acade6a4c8f82ae5c4cc787c2a65fae26a731","instance_id":"068abb965aa848178174ca1fdb56864e","private_instance_id":"5d948dbeb7a042a4a8478a9359c896c1a3b2e76a9ade4b2a8d81c1f35a2dfe7d","state":"CRASHED","state_timestamp":1433973049.7428112,"state_born_timestamp":1433973047.9089131,"state_starting_timestamp":1433973047.9103467,"warden_handle":"18o62fv0gjs","state_crashed_timestamp":1433973049.7428186},"duration":1.832755862,"error":"Script exited with status 2","backtrace":["/var/vcap/packages/dea_next/lib/container/container.rb:81:in `run_script'","/var/vcap/packages/dea_next/lib/dea/starting/instance.rb:427:in `block in promise_extract_droplet'","/var/vcap/packages/dea_next/lib/dea/promise.rb:92:in `call'","/var/vcap/packages/dea_next/lib/dea/promise.rb:92:in `block in run'"]},"thread_id":70101165527800,"fiber_id":70101199036880,"process_id":28791,"file":"/var/vcap/packages/dea_next/lib/dea/task.rb","lineno":104,"method":"block in resolve_and_log"}
```

From warden:
```json
{"timestamp":1433973460.4345205,"message":"link (took 0.620225)","log_level":"debug","source":"Warden::Container::Linux","data":{"handle":"18o62fv0glc","request":{"handle":"18o62fv0glc","job_id":520},"response":{"exit_status":2,"stdout":"","stderr":"tar: ./logs: Cannot mkdir: Permission denied\ntar: ./logs: Cannot mkdir: Permission denied\ntar: ./logs/staging_task.log: Cannot open: No such file or directory\ntar: ./tmp: Cannot mkdir: Permission denied\ntar: ./staging_info.yml: Cannot open: Permission denied\ntar: .: Cannot utime: Operation not permitted\ntar: Exiting with failure status due to previous errors\n","info":"#<Warden::Protocol::InfoResponse:0x007f4e9ade0040>"}},"thread_id":69987779629820,"fiber_id":69987791368560,"process_id":28652,"file":"/var/vcap/data/packages/warden/d9a748943436336f174ba060f11391ba6dae6f97.1-29b2665da7860900c7c33ae877ac11526427dc01/warden/lib/warden/container/base.rb","lineno":300,"method":"dispatch"}
{"timestamp":1433973460.471933,"message":"run (took 0.662887)","log_level":"debug","source":"Warden::Container::Linux","data":{"handle":"18o62fv0glc","request":{"handle":"18o62fv0glc","script":"cd /home/vcap/ && tar zxf /var/vcap/data/dea_next/droplets/d34b00e58cf56256d8fa6c2ea62f73f7ae9c3f8a/droplet.tgz"},"response":{"exit_status":2,"stdout":"","stderr":"tar: ./logs: Cannot mkdir: Permission denied\ntar: ./logs: Cannot mkdir: Permission denied\ntar: ./logs/staging_task.log: Cannot open: No such file or directory\ntar: ./tmp: Cannot mkdir: Permission denied\ntar: ./staging_info.yml: Cannot open: Permission denied\ntar: .: Cannot utime: Operation not permitted\ntar: Exiting with failure status due to previous errors\n","info":"#<Warden::Protocol::InfoResponse:0x007f4e9ad98b28>"}},"thread_id":69987779629820,"fiber_id":69987791368560,"process_id":28652,"file":"/var/vcap/data/packages/warden/d9a748943436336f174ba060f11391ba6dae6f97.1-29b2665da7860900c7c33ae877ac11526427dc01/warden/lib/warden/container/base.rb","lineno":300,"method":"dispatch"}
{"timestamp":1433973460.589916,"message":"Exited with status 23 (0.111s): [[\"/var/vcap/data/packages/warden/d9a748943436336f174ba060f11391ba6dae6f97.1-29b2665da7860900c7c33ae877ac11526427dc01/warden/src/closefds/closefds\", \"/var/vcap/data/packages/warden/d9a748943436336f174ba060f11391ba6dae6f97.1-29b2665da7860900c7c33ae877ac11526427dc01/warden/src/closefds/closefds\"], \"rsync\", \"-e\", \"/var/vcap/data/warden/depot/18o62fv0glc/bin/wsh --socket /var/vcap/data/warden/depot/18o62fv0glc/run/wshd.sock --rsh\", \"-r\", \"-p\", \"--links\", \"vcap@container:/home/vcap/logs/\", \"/var/vcap/data/dea_next/crashes/2307a7452a474311948d5704c3bb122a/logs\"]","log_level":"warn","source":"Warden::Container::Linux","data":{"handle":"18o62fv0glc","stdout":"","stderr":"rsync: change_dir \"/home/vcap/logs\" failed: No such file or directory (2)\nrsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1655) [Receiver=3.1.0]\nrsync: [Receiver] write error: Broken pipe (32)\n"},"thread_id":69987779629820,"fiber_id":69987792826240,"process_id":28652,"file":"/var/vcap/data/packages/warden/d9a748943436336f174ba060f11391ba6dae6f97.1-29b2665da7860900c7c33ae877ac11526427dc01/warden/lib/warden/container/spawn.rb","lineno":135,"method":"set_deferred_success"}
```

It appears as though the vcap user created during the docker build is conflicting with warden. [We have generated the rootfs in question from the cflinuxfs2 docker file on the add-vcap-user branch of the stacks repo](https://github.com/cloudfoundry/stacks/blob/add-vcap-user/cflinuxfs2/Dockerfile). We believe that the adduser command is correctly creating the home directory and adding the user to `/etc/passwd` but warden still cannot access that.

As far as we can tell the stack trace for the DEA at this point starts at [`promise_extract_droplet` in instance.rb:423](https://github.com/cloudfoundry/dea_ng/blob/master/lib/dea/starting/instance.rb#L423)

**Comment by svett** [comment]
Hi @zaksoup,

We will investigate this and we will get back to you soon.

Thanks!

**Comment by svett** [comment]
@jtarchie We are going to send a PR to cf-release to move the extraction of the `cflinuxfs2 rootfs` into the `dea_next` job vm. This will ensure that the ownership of `/home/vcap` is `vcap:vcap`.

The compilation & distribution steps do not preserve the ownership and that's is why we are moving the extraction of the rootfs to the ctl script.
